File: android/app/src/test/java/com/remotedisplay/player/trigger/TriggerResolveTest.kt
package com.remotedisplay.player.trigger

import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * Drift guard: the Kotlin resolver must agree with the SHARED contract at
 * shared/trigger-vectors.json — the same file the JS player's suite is held to. No snapshot is
 * taken: the test task points `triggerVectors` at the single source (see app/build.gradle.kts), so
 * any TriggerResolve.kt change that breaks a vector fails CI.
 *
 * ⚠️ Why this is worth a whole file: the fire path decides whether an unauthenticated packet from a
 * LAN changes what is on a screen. Two implementations of that in two languages drift silently, and
 * the drift is security-relevant — an Android panel accepting a payload the web player refuses is a
 * hole nobody would notice until it was used. Same reasoning, same mechanism, as ScheduleEvalTest.
 */
class TriggerResolveTest {

    private fun triggersFrom(arr: com.google.gson.JsonArray): List<TriggerResolve.Trigger> =
        arr.map { el ->
            val o = el.asJsonObject
            TriggerResolve.Trigger(
                id = o.get("id").asString,
                name = o.get("name").asString,
                matchToken = o.get("match_token").asString,
                clearToken = if (o.get("clear_token").isJsonNull) null else o.get("clear_token").asString,
                sourceHttp = o.get("source_http").asBoolean,
                sourceUdp = o.get("source_udp").asBoolean,
                mode = o.get("mode").asString,
                priority = if (o.has("priority")) o.get("priority").asInt else 0
            )
        }

    @Test
    fun conformsToSharedVectors() {
        val path = System.getProperty("triggerVectors")
            ?: error("triggerVectors system property not set (configured in app/build.gradle.kts)")
        val root = JsonParser.parseString(File(path).readText()).asJsonObject

        val defaultTriggers = triggersFrom(root.getAsJsonArray("triggers"))
        val defaultSecret = root.get("device_secret").asString
        val defaultClearAll =
            if (root.get("clear_all_token").isJsonNull) null else root.get("clear_all_token").asString

        val failures = StringBuilder()
        var count = 0

        for (el in root.getAsJsonArray("vectors")) {
            val v = el.asJsonObject
            val desc = v.get("description").asString

            // Per-vector overrides, so a vector can describe an unprovisioned device or one holding
            // no triggers without needing a second file.
            val triggers = if (v.has("triggers")) triggersFrom(v.getAsJsonArray("triggers")) else defaultTriggers
            val secret = if (v.has("device_secret")) v.get("device_secret").asString else defaultSecret
            val clearAll = when {
                !v.has("clear_all_token") -> defaultClearAll
                v.get("clear_all_token").isJsonNull -> null
                else -> v.get("clear_all_token").asString
            }

            val got = TriggerResolve.evaluate(
                text = v.get("text").asString,
                triggers = triggers,
                deviceSecret = secret,
                clearAllToken = clearAll,
                source = v.get("source").asString
            )

            val expect = v.getAsJsonObject("expect")
            val wantOk = expect.get("ok").asBoolean
            if (got.ok != wantOk) {
                failures.append("\n  [$desc] ok: expected $wantOk got ${got.ok} (reason=${got.reason})")
            } else if (wantOk) {
                val wantAction = expect.get("action").asString
                if (got.action.toString() != wantAction) {
                    failures.append("\n  [$desc] action: expected $wantAction got ${got.action}")
                }
                val wantId = if (expect.get("trigger_id").isJsonNull) null else expect.get("trigger_id").asString
                if (got.trigger?.id != wantId) {
                    failures.append("\n  [$desc] trigger: expected $wantId got ${got.trigger?.id}")
                }
            } else {
                val wantReason = expect.get("reason").asString
                if (got.reason.toString() != wantReason) {
                    failures.append("\n  [$desc] reason: expected $wantReason got ${got.reason}")
                }
            }
            count++
        }

        if (failures.isNotEmpty()) {
            error("Kotlin resolver disagrees with shared/trigger-vectors.json:$failures\n" +
                  "The vectors are the contract. If this fails, this implementation is wrong.")
        }
        // A vector file that silently emptied would pass every assertion above.
        assertTrue("only $count vectors ran — the contract file looks truncated", count >= 25)
    }

    @Test
    fun secretCompareIsLengthChecked() {
        assertEquals(true, TriggerResolve.secretMatches("abc", "abc"))
        assertEquals(false, TriggerResolve.secretMatches("abc", "abcd"))
        assertEquals(false, TriggerResolve.secretMatches(null, "abc"))
    }
}

File: server/test/trigger-resolve.test.js
'use strict';

/*
 * The decision half of the fire path. docs/triggers-design.md §2, §11.
 *
 * ⚠️ This module decides whether an unauthenticated packet arriving from a LAN changes what is on a
 * screen. The tests that matter are the refusals, not the happy path.
 */
const { test } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const TR = require('../lib/trigger-resolve');

const SECRET = 'a'.repeat(32);
const TRIGGERS = [
  { id: 't1', name: 'Evac', match_token: 'EVAC', clear_token: 'EVAC_CLR',
    source_http: true, source_udp: true, mode: 'until_cleared' },
  { id: 't2', name: 'UDP only', match_token: 'UDPONLY', clear_token: null,
    source_http: false, source_udp: true, mode: 'once' },
  { id: 't3', name: 'HTTP only', match_token: 'HTTPONLY', clear_token: null,
    source_http: true, source_udp: false, mode: 'once' },
];
const ev = (text, over = {}) => TR.evaluate({
  text, triggers: TRIGGERS, deviceSecret: SECRET, source: 'udp', ...over });

test('a well-formed fire resolves to its trigger', () => {
  const r = ev(`ST1 ${SECRET} EVAC`);
  assert.equal(r.ok, true);
  assert.equal(r.action, 'fire');
  assert.equal(r.trigger.id, 't1');
});

test('a clear token resolves to clear, on the same trigger', () => {
  const r = ev(`ST1 ${SECRET} EVAC_CLR`);
  assert.equal(r.action, 'clear');
  assert.equal(r.trigger.id, 't1');
});

test('a device-level clear-all resolves without naming a trigger', () => {
  const r = ev(`ST1 ${SECRET} ALLSTOP`, { clearAllToken: 'ALLSTOP' });
  assert.equal(r.action, 'clear_all');
  assert.equal(r.trigger, undefined);
});

test('a trailing newline is tolerated — most gear appends one', () => {
  assert.equal(ev(`ST1 ${SECRET} EVAC\n`).action, 'fire');
  assert.equal(ev(`ST1 ${SECRET} EVAC\r\n`).action, 'fire');
});

// ───────────────────────── refusals ─────────────────────────

test('⚠️ the wrong secret is refused, and the reason says so', () => {
  const r = ev(`ST1 ${'b'.repeat(32)} EVAC`);
  assert.equal(r.ok, false);
  assert.equal(r.reason, 'bad_secret');
});

test('⚠️ a token valid on ANOTHER screen resolves to nothing here', () => {
  // Scoping happened at sync time: this device only holds its own assigned triggers, so a token
  // that is perfectly real elsewhere is simply unknown. No cross-screen firing.
  const r = TR.evaluate({ text: `ST1 ${SECRET} SOMEONE_ELSES`, triggers: TRIGGERS,
    deviceSecret: SECRET, source: 'udp' });
  assert.equal(r.reason, 'unknown_token');
});

test('⚠️ a UDP-only trigger cannot be fired over HTTP', () => {
  // Enabling only UDP says something specific: fired by the panel wired to the alarm, not by
  // anything that can reach the box over HTTP. The device-level flag must not widen that.
  assert.equal(ev('ST1 ' + SECRET + ' UDPONLY', { source: 'http' }).reason, 'unknown_token');
  assert.equal(ev('ST1 ' + SECRET + ' UDPONLY', { source: 'udp' }).action, 'fire');
});

test('⚠️ an HTTP-only trigger cannot be fired over UDP', () => {
  assert.equal(ev('ST1 ' + SECRET + ' HTTPONLY', { source: 'udp' }).reason, 'unknown_token');
  assert.equal(ev('ST1 ' + SECRET + ' HTTPONLY', { source: 'http' }).action, 'fire');
});

test('broadcast noise is rejected on the magic, before any parsing', () => {
  // On subnet broadcast this socket sees mDNS, discovery chatter, a printer announcing itself.
  for (const junk of ['', 'hello', '{"jsonrpc":"2.0"}', 'M-SEARCH * HTTP/1.1']) {
    assert.equal(TR.parseWire(junk).reason, junk === '' ? 'bad_magic' : 'bad_magic');
  }
});

test('an oversized payload is refused before work is done on it', () => {
  const r = TR.parseWire('ST1 ' + SECRET + ' ' + 'x'.repeat(600));
  assert.equal(r.reason, 'too_large');
});

test('a token containing a space cannot arrive — the field count refuses it', () => {
  // The editor refuses this at save time; this is the second line of defence, because the wire
  // format cannot represent it either way.
  assert.equal(TR.parseWire(`ST1 ${SECRET} FIRE ALARM`).reason, 'malformed');
});

test('a missing secret field is malformed, not a bad secret', () => {
  // The distinction matters for the counters: malformed means the sender's format is wrong,
  // bad_secret means the format is right and the credential is not.
  assert.equal(TR.parseWire('ST1 EVAC').reason, 'malformed');
});

test('no configured device secret refuses everything', () => {
  // An unprovisioned device must be inert rather than open.
  const r = TR.evaluate({ text: `ST1 ${SECRET} EVAC`, triggers: TRIGGERS, deviceSecret: '', source: 'udp' });
  assert.equal(r.reason, 'bad_secret');
});

test('secret compare is length-checked', () => {
  assert.equal(TR.secretMatches('abc', 'abc'), true);
  assert.equal(TR.secretMatches('abc', 'abcd'), false);
  assert.equal(TR.secretMatches('', ''), true);
  assert.equal(TR.secretMatches(null, 'abc'), false);
});

// ───────────────────────── rate limiting ─────────────────────────

test('a burst is allowed, then throttled, then recovers', () => {
  const rl = TR.createRateLimiter({ perSec: 5, burst: 3, globalPerSec: 1000 });
  let now = 0;
  assert.equal(rl.allow('10.0.0.1', now), true);
  assert.equal(rl.allow('10.0.0.1', now), true);
  assert.equal(rl.allow('10.0.0.1', now), true);
  assert.equal(rl.allow('10.0.0.1', now), false, 'burst exhausted');
  now += 1000;                                   // one second of refill at 5/s
  assert.equal(rl.allow('10.0.0.1', now), true);
});

test('⚠️ one noisy source cannot starve another', () => {
  const rl = TR.createRateLimiter({ perSec: 1, burst: 2, globalPerSec: 1000 });
  rl.allow('10.0.0.1', 0); rl.allow('10.0.0.1', 0);
  assert.equal(rl.allow('10.0.0.1', 0), false);
  assert.equal(rl.allow('10.0.0.2', 0), true, 'a second source has its own bucket');
});

test('⚠️ a global ceiling stops a spoofed-source flood walking around the per-source limit', () => {
  const rl = TR.createRateLimiter({ perSec: 100, burst: 100, globalPerSec: 3 });
  let allowed = 0;
  for (let i = 0; i < 50; i++) if (rl.allow('10.0.0.' + i, 0)) allowed++;
  assert.equal(allowed, 3, 'each packet from a fresh source would otherwise get a fresh bucket');
});

test('the bucket map is pruned, so a flood is not a slow leak', () => {
  const rl = TR.createRateLimiter({ globalPerSec: 1000 });
  for (let i = 0; i < 20; i++) rl.allow('10.0.0.' + i, 0);
  assert.equal(rl.size, 20);
  rl.prune(400000);
  assert.equal(rl.size, 0);
});

// ───────────────────────── the UMD trap ─────────────────────────

test('⚠️ the module exports BOTH ways, not as an if/else', () => {
  /*
   * A node-enabled BrightSign widget is a browser and a CommonJS context at once. A UMD that tests
   * module.exports first and falls back to the global takes the CommonJS branch there, so
   * window.TriggerResolve is never assigned and every call site silently sees undefined. That has
   * already happened here to transitions, dayparting, mute and wall geometry.
   */
  const src = require('fs').readFileSync(require.resolve('../lib/trigger-resolve.js'), 'utf8');
  const tail = src.slice(src.indexOf('if (typeof module'));
  assert.doesNotMatch(tail, /\belse\b/,
    'an else here means the window global is skipped on a node-enabled widget');
  assert.match(tail, /window\.TriggerResolve = API/);
  assert.match(tail, /module\.exports = API/);
});

/*
 * ⚠️ THE SHARED CONTRACT. shared/trigger-vectors.json is consumed by this suite AND by the Kotlin
 * player's TriggerResolveTest. Two implementations of a fire path in two languages WILL drift, and
 * the drift is silent and security-relevant: one player accepting a payload another refuses is a
 * hole nobody sees until it is used. The vectors are the artifact; the implementations are what is
 * being checked. Same pattern as shared/schedule-vectors.json.
 */
test('⚠️ every shared vector holds for the JS implementation', () => {
  const vectors = JSON.parse(fs.readFileSync(
    path.join(__dirname, '..', '..', 'shared', 'trigger-vectors.json'), 'utf8'));

  let checked = 0;
  for (const v of vectors.vectors) {
    const got = TR.evaluate({
      text: v.text,
      triggers: v.triggers !== undefined ? v.triggers : vectors.triggers,
      deviceSecret: v.device_secret !== undefined ? v.device_secret : vectors.device_secret,
      clearAllToken: v.clear_all_token !== undefined ? v.clear_all_token : vectors.clear_all_token,
      source: v.source,
    });
    const where = `vector: ${v.description}`;
    assert.equal(got.ok, v.expect.ok, where);
    if (v.expect.ok) {
      assert.equal(got.action, v.expect.action, where);
      assert.equal(got.trigger ? got.trigger.id : null, v.expect.trigger_id, where);
    } else {
      assert.equal(got.reason, v.expect.reason, where);
    }
    checked++;
  }
  // A vector file that silently emptied would pass every assertion above.
  assert.ok(checked >= 25, `only ${checked} vectors ran — the contract file looks truncated`);
});
